learning-from-errors/overlapping-instances/article.md

```diff
@@ -0,0 +1,626 @@
+
+## Flip-Flopping overlapping instances
+
+Here we look at a variant of this error that seemingly flip-flops when
+attempts are made to fix it. Let us look at a sample.
+
+```hs
+{-# LANGUAGE FlexibleInstances #-}
+
+module Main (main) where
+
+import Data.Typeable
+
+class Printable a where
+  printMe :: a -> IO ()
+
+instance Printable a where
+  printMe _ = putStrLn "General instance"
+
+instance Functor f => Printable (f a) where
+  printMe _ = putStrLn "Instance for a Functor"
+
+newtype MyType a = MyType a
+
+main :: IO ()
+main = printMe (MyType 5)
+
+```
+
+As expected, we get the overlapping instances error.
+
+```
+• Overlapping instances for Printable (MyType Char)
+    arising from a use of ‘printMe’
+  Matching instances:
+    instance Printable a -- Defined at app/Main.hs:10:10
+    instance Functor f => Printable (f a)
+      -- Defined at app/Main.hs:13:10
+```
+
+Since we saw a similar error in the last example and fixed it by
+removing one of the instances, it might appear that same could work here as well.
+
+And it also kind of makes sense. Since GHC is confused by two eligible instances, in all probability, removing one of them should fix it, right?
+
+So we try by removing the instance for `Printable a`.
+
+<details>
+  <summary>Show changed code</summary>
+
+```hs
+{-# LANGUAGE FlexibleInstances #-}
+
+module Main (main) where
+
+import Data.Typeable
+
+class Printable a where
+  printMe :: a -> IO ()
+
+instance Functor f => Printable (f a) where
+  printMe _ = putStrLn "Instance for a Functor"
+
+newtype MyType a = MyType a
+
+main :: IO ()
+main = printMe (MyType 'c')
+```
+</details>
+
+And when we re-compile, we get...
+
+```
+• No instance for (Functor MyType) arising from a use of ‘printMe’
+• In the expression: printMe (MyType 'c')
+  In an equation for ‘main’: main = printMe (MyType 'c')
+```
+
+To our great surprise, we find that removing one instance from the two eligible
+instances made GHC reject the remaining instance as well!
+
+So we end up in the situation where GHC flip-flops between these two errors.
+
+It shows that it might not be a good idea to blindly remove instances when you
+come across overlapping instance errors. It is better to carefully examine the
+actual cause of the issue, keeping in mind the algorithm GHC follows to resolve
+instances.
+
+Let us walk through this algorithm and see why the first error happens.
+
+So the very first step is:
+
+```
+Find all instances I that match the target constraint; that is, the target
+constraint is a substitution instance of I. These instance declarations are the
+candidates.
+```
+
+Our target constraint here is `MyType Char`, and both the instances for `Printable a`
+and `Printable (f a)` match.
+
+The next step says:
+
+```
+If no candidates remain, the search fails.
+```
+
+Since we have two instances that match, we can continue to the next step, which says,
+
+```
+Eliminate any candidate `IX` for which there is another candidate `IY` such that
```

Comment:
Likewise, this is text, not code, so it shouldn’t be monospace. And `IX` and `IY` must be typeset as math.